src/Microsoft.ML.Tokenizers/PreTokenizer/TikTokenPreTokenizer.cs

```diff
@@ -0,0 +1,176 @@
+using System.Diagnostics;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.ML.Tokenizers
+{
+    /// <summary>
+    /// The pre-tokenizer for Tiktoken tokenizer.
+    /// </summary>
+    public sealed class TikTokenPreTokenizer : PreTokenizer
+    {
+        private readonly Regex? _specialTokensRegex;
+        private readonly Regex _regex;
+
+        public TikTokenPreTokenizer(string regexPattern, IReadOnlyDictionary<string, int>? specialTokensEncoder)
+        {
+            if (regexPattern is null)
+            {
+                throw new ArgumentNullException(nameof(regexPattern));
+            }
+
+            _regex = new Regex(regexPattern, RegexOptions.Compiled);
+
+            if (specialTokensEncoder is not null && specialTokensEncoder.Count > 0)
+            {
+                _specialTokensRegex = new Regex(string.Join("|", specialTokensEncoder.Keys.Select(s => Regex.Escape(s))), RegexOptions.Compiled);
+            }
+        }
+
+        /// <summary>
+        /// Splits the given string in multiple substrings at the word boundary, keeping track of the offsets of said substrings from the original string.
+        /// </summary>
+        /// <param name="sentence">The string to split into tokens.</param>
+        /// <param name="skipSpecialTokens">Indicates whether to skip the special tokens.</param>
+        /// <returns>The list of the splits containing the tokens and the token's offsets to the original string.</returns>
+        public override IEnumerable<Split> PreTokenize(string sentence, bool skipSpecialTokens = false)
+        {
+            if (string.IsNullOrEmpty(sentence))
+            {
+                return EmptyList;
+            }
+
+            return new TokenizationEnumerable(sentence, _regex, skipSpecialTokens ? null : _specialTokensRegex);
+        }
+
+        private readonly struct TokenizationEnumerable : IEnumerable<Split>
+        {
+            private readonly string _sentence;
+            private readonly Regex _regex;
+            private readonly Regex? _specialTokensRegex;
+
+            public TokenizationEnumerable(string sentence, Regex regex, Regex? specialTokensRegex)
+            {
+                if (sentence is null)
+                {
+                    throw new ArgumentNullException(nameof(sentence));
+                }
+
+                if (regex is null)
+                {
+                    throw new ArgumentNullException(nameof(regex));
+                }
+
+                _sentence = sentence;
+                _regex = regex;
+                _specialTokensRegex = specialTokensRegex;
+            }
+
+            public readonly IEnumerator<Split> GetEnumerator() => new TokenizationEnumerator(_sentence, _regex, _specialTokensRegex);
+            IEnumerator IEnumerable.GetEnumerator() => new TokenizationEnumerator(_sentence, _regex, _specialTokensRegex);
+
+            private struct TokenizationEnumerator : IEnumerator<Split>
```

Comment:
All of these enumerables/enumerators being structs only makes sense if the compiler consuming them can statically see their types. If they're just immediately returned as the interfaces, they should just be classes.